server/src/main/java/org/elasticsearch/index/codec/bloomfilter/ES87BloomFilterPostingsFormat.java

```diff
@@ -0,0 +1,1174 @@
+                    return getDelegate().termState();
+                }
+            };
+        }
+    }
+
+    private abstract static class LazyFilterTermsEnum extends BaseTermsEnum {
+        abstract TermsEnum getDelegate() throws IOException;
+
+        @Override
+        public SeekStatus seekCeil(BytesRef text) throws IOException {
+            return getDelegate().seekCeil(text);
+        }
+
+        @Override
+        public void seekExact(long ord) throws IOException {
+            getDelegate().seekExact(ord);
+        }
+
+        @Override
+        public BytesRef term() throws IOException {
+            return getDelegate().term();
+        }
+
+        @Override
+        public long ord() throws IOException {
+            return getDelegate().ord();
+        }
+
+        @Override
+        public int docFreq() throws IOException {
+            return getDelegate().docFreq();
+        }
+
+        @Override
+        public long totalTermFreq() throws IOException {
+            return getDelegate().totalTermFreq();
+        }
+
+        @Override
+        public PostingsEnum postings(PostingsEnum reuse, int flags) throws IOException {
+            return getDelegate().postings(reuse, flags);
+        }
+
+        @Override
+        public ImpactsEnum impacts(int flags) throws IOException {
+            return getDelegate().impacts(flags);
+        }
+
+        @Override
+        public BytesRef next() throws IOException {
+            return getDelegate().next();
+        }
+
+        @Override
+        public AttributeSource attributes() {
+            try {
+                return getDelegate().attributes();
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        }
+    }
+
+    static int bloomFilterSize(int maxDocs) {
+        if (maxDocs < 1) {
+            throw new IllegalStateException("maxDocs must be greater than or equal to 1, got " + maxDocs);
+        }
+        // 10% saturation (i.e., 10 bits for each term)
+        long numBits = maxDocs * (long) BITS_PER_ENTRY;
+        // Round to the next number of 8 since we can only store whole bytes
+        numBits = ((numBits - 1) | 0x07L) + 1;
+        if (numBits > Integer.MAX_VALUE) {
+            return Integer.MAX_VALUE;
+        } else {
+            return (int) numBits;
+        }
+    }
+
+    static int numBytesForBloomFilter(int bloomFilterSize) {
+        return Math.toIntExact((bloomFilterSize + 7L) / 8L);
+    }
+
+    // Uses CityHash64 to generate a 64-bit hash value, then picks 7 subsets of 31 bits each and returns the values in the
+    // outputs array. This provides us with 7 reasonably independent hashes of the data for the cost of one CityHash64 calculation.
+    static int[] hashTerm(BytesRef br, int[] outputs) {
+        final long hash_v2 = CityHash.cityHash64(br.bytes, br.offset, br.length);
+        final int upper_half = (int) hash_v2 >> 32;
```

Comment:
IIRC the cast has priority over the shift, and since shifts are modulo 32 on integers, you'd get the same value for `upper_half` and `lower_half`. This should be `(int) (hash_v2 >> 32)`.

Reply:
Good catch, thank you!